Add advanced tips section (#268)
eclipse-theia/theia-website: src/docs/tips.md

---
title: Advanced Tips
---

# Advanced Tips

In this section we'll outline some advanced hints and tips to get the most out of developing with Eclipse Theia.

## Upgrading functionality in a VS Code Extension for use in Theia

> It's possible to declare a custom Theia API namespace which can be implemented in a standard VS Code Extension.

Reviewer: Theia allows running VS Code extension by providing a compatible API (see [this overview](https://theia-ide.org/docs/extensions/) for details). It is possible to extend this API to allow VS Code extensions running in Theia to access additional functionality compared to when they run within VS Code.

> In this way, common functionality can be written to target VS Code and Theia, with enhancements appearing just in the Theia IDE.

Reviewer: ```suggestion
This allows you to provide a feature as a VS Code extension targeting VS Code and Theia. However, when running in Theia, the feature can be enhanced by using custom API only available in Theia.
```

> Usage of the custom API is invoked only when run in a Theia IDE, determined via the application name.

Reviewer: ```suggestion
The following code example shows the usage custom API which is invoked only when running in a Theia-based application. This is determined via the application name.
```

The API is imported asynchronously to avoid runtime errors in VS Code.

```typescript

> if (vscode.env.appName === APP_NAME) {

Reviewer: ```suggestion
if (vscode.env.appName === MY_THEIA_APP_NAME) {
```

    // Implement Theia API

>     const api = await import('@extension/api');

Reviewer: ```suggestion
    const api = await import('@mytheiaextension/mycustomapi');
```

Reviewer: Is this correct? If so, I think it would be more understandable.

>     api.host.getMessageHandler(() => getMessage());

Reviewer: ```suggestion
    // After importing the custom API, it can be used as any other API. The following lines are using an example API.
    api.host.getMessageHandler(() => getMessage());
```

    api.host.onRequestMessage((actor: string) => {
        const message = getMessage(actor);
        api.host.showMessage(message);
    });

>     // Execute Theia custom command

Reviewer: ```suggestion
}
```
Alternatively to providing a custom API, you can also define custom commands. Again, these commands would only be available if the VS Code extension is running in Theia (see following code example).
    ```typescript
    if (vscode.env.appName === MY_THEIA_APP_NAME) {
    // Execute Theia custom command
```

    const commands = await vscode.commands.getCommands();
    if (commands.indexOf(THEIA_CUSTOM_COMMAND) > -1) {
        vscode.commands.executeCommand(THEIA_CUSTOM_COMMAND);
    }
}
```

> This works due to the way the API support is injected by the Theia plugin host.

Reviewer: Would still remove this, do not know what exactly we are telling the reader here.

> }

Author: fixed
